NN model based on the docs

KaspervanM opened the pull request:
This adds a simple model based on the pytorch docs. It includes a custom dataset using local images in some folder.

GPU is not yet supported and the application doesn't work yet. For that, the problem and fix are known, just not yet implemented (see bottom of NN.py)

Detaching tensors should NOT be done if you want to backpropagate through them.
[source I found it on](https://discuss.pytorch.org/t/element-0-of-tensors-does-not-require-grad-and-does-not-have-a-grad-fn/32908/28)
[source Copilot generated](https://discuss.pytorch.org/t/why-do-we-need-to-detach-variable-which-contains-hidden-representation/1426/2)
fix: implement the functions in python using Tensor operations. 

The functions are:
- the blur
- the normalization
- polarization (i.e., denormalization)

KaspervanM committed:
feat: NN model based on the docs.

GPU is not yet supported and the problem and fix are known, just not yet implemented (see bottom of NN.py)

KaspervanM committed:
fix: added source in ImageDataset.py

Signed-off-by: KaspervanM <kaspervanm@gmail.com>

evanpacini commented:
This all sounds like it could work if using the tensor operations indeed. Quite insane if pytorch is able to continue the gradient calculations within our blur function.

KaspervanM committed:
fix: model now working and trainable with more data available

Signed-off-by: KaspervanM <kaspervanm@gmail.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by black

restyled-commits committed:
Restyled by isort

restyled-commits committed:
Restyled by reorder-python-imports

restyled-commits committed:
Restyled by yapf

KaspervanM commented:
Fixed the issues and now a trainable model. I added more training data, I but could not find any improvements yet. This will probably cost too much computing power to do in reasonable time on my laptop.

At the top of NN.py, there are a bunch of parameters you can play around with.

Interrupting the training by a keyboard interrupt, will save the model before shutting down. Test a model by setting `SHOULD_TRAIN = False` and give the correct `LOAD_MODEL_NAME`.

KaspervanM committed:
Merge pull request #3 from evanpacini/restyled/some_NN_attempts

Restyle NN model based on the docs

KaspervanM committed:
feat: writing images now supported by the interface.

Signed-off-by: KaspervanM <kaspervanm@gmail.com>

KaspervanM committed:
feat: more images and writing images

Signed-off-by: KaspervanM <kaspervanm@gmail.com>
